### ISSUE TITLE
Model.dict() is taking to much time to respond

### ISSUE THREAD

**Issue by luisjimenez6245** [collerek/ormar#337]
**Describe the bug**
Im using a model with 1 many to many rel and 1 foreign key, with ormar, but the seralization is taking too much to do it. After disabling dict() the overloaded method, the response model does not have any related model, but it takes almost a 1/100  (I didn't make the exact calculations) of the time to do it. This problem also affects  to make the json response, so we should expect a slow response time. Is there any way to disable this override method or is there any plan/idea to optimize it? 

With ormar dict():
<img width="453" alt="Captura de Pantalla 2021-09-06 a la(s) 12 29 23" src="https://user-images.githubusercontent.com/26035464/132249156-d4553924-3c82-48c9-b57b-b1489004538e.png">

With only-pydantic dict():
<img width="410" alt="Captura de Pantalla 2021-09-06 a la(s) 12 28 13" src="https://user-images.githubusercontent.com/26035464/132249158-708ca324-e146-4fa8-8f6e-5f735cdaf969.png">

The list of models that I´m loading doesn't have any related model loaded, neither items on the many to many relation

**Comment by collerek** [comment]
I'm trying to optimize the `dict()` if I can on a regular basis but to check your scenario I would need your models and sample data to make some tests.

That seems like a quite long response indeed, do you load the related models in the query or not?

Because if you do, then you serialize the main model plus n x related model, where n is a number of related model instances of the main model. So you are comparing the time of serializing one model (pydantic) versus n+1 models (ormar).

Do you want the related model included in the dict or not?

**Comment by luisjimenez6245** [comment]
In answer to your questions, im not loading the related models.

Here are my principal models that im working on:


```python
class Release(orm.Model):

    id: int = orm.Integer(name="id", primary_key=True)
    title: str = orm.String(max_length=200, default='')
    description: str = orm.Text(default='')
    tag: Tag = orm.ForeignKey(Tag, ondelete='CASCADE', onupdate='CASCADE')
    changelogs:  List[Changelog] = orm.ManyToMany(
        Changelog, through=ChagenlogRelease, ondelete='CASCADE', onupdate='CASCADE'
    )
    data: dict = orm.JSON(default={})

class Project(orm.Model):

    id: int = orm.Integer(name="id", primary_key=True)
    name: str = orm.String(max_length=100)
    description: str = orm.Text(default='')
    git_url: str = orm.String(max_length=500, default='')
    labels: Optional[Union[List[Label], Label]] = orm.ManyToMany(
        Label, through=ProjectLabel, ondelete='CASCADE', onupdate='CASCADE'
    )
    changelog_jira_tag: str = orm.String(max_length=100, default='')
    change_type_jira_tag: str = orm.String(max_length=100, default='')
    jira_prefix: str = orm.String(max_length=10, default='SAN')
    type: str = orm.String(max_length=10, default='cs')
    target_branch_name: str = orm.String(max_length=100, default='master')
    header: str = orm.String(max_length=250, default='')
    jira_url: str = orm.String(
        max_length=500,
    )
    changelog_file: str = orm.String(max_length=250, default='')
    version_file: str = orm.String(max_length=250, default='')
    members: Optional[Union[List[User], User]] = orm.ManyToMany(
        User, through=UserProject, ondelete='CASCADE', onupdate='CASCADE'
    )
```

The response I´m getting with pydantic .dict() func
```json
[
    {
        "id": 9,
        "title": "prueba-2321",
        "description": "\n<!--- start changelog ver.v.1.3.0.0 -->\n### [v.1.3.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.3.0.0 -->\n",
        "data": {}
    },
    {
        "id": 8,
        "title": "prueba-123-prod",
        "description": "\n<!--- start changelog ver.v.1.2.0.0 -->\n### [v.1.2.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.2.0.0 -->\n",
        "data": {}
    },
    {
        "id": 6,
        "title": "prueba-3-2",
        "description": "\n<!--- start changelog ver.v.1.1.0.0 -->\n### [v.1.1.0.0] - 2021-07-29\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.1.0.0 -->\n",
        "data": {}
    }
]
```

The response I´m getting with ormar .dict() func
```json
[
    {
        "id": 9,
        "title": "prueba-2321",
        "description": "\n<!--- start changelog ver.v.1.3.0.0 -->\n### [v.1.3.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.3.0.0 -->\n",
        "data": {},
        "tag": {
            "id": 18,
            "taglabel": null,
            "tagcommit": null,
            "tagissue": null
        },
        "changelogs": []
    },
    {
        "id": 8,
        "title": "prueba-123-prod",
        "description": "\n<!--- start changelog ver.v.1.2.0.0 -->\n### [v.1.2.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.2.0.0 -->\n",
        "data": {},
        "tag": {
            "id": 17,
            "taglabel": null,
            "tagcommit": null,
            "tagissue": null
        },
        "changelogs": []
    },
    {
        "id": 6,
        "title": "prueba-3-2",
        "description": "\n<!--- start changelog ver.v.1.1.0.0 -->\n### [v.1.1.0.0] - 2021-07-29\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.1.0.0 -->\n",
        "data": {},
        "tag": {
            "id": 12,
            "taglabel": null,
            "tagcommit": null,
            "tagissue": null
        },
        "changelogs": []
    }
]
```

**Comment by collerek** [comment]
Do you mean you get 52s with those 3 models in response or is it just a sample structure?

**Comment by luisjimenez6245** [comment]
Im getting the 52s in that exact response :((

**Comment by collerek** [comment]
Ok there is something seriously wrong here, the whole test suite of ormar runs under 40s on all backends and it's over 380 tests with queries a lot more nested than yours. Will try to check that. 

So you use the latest version of ormar?

**Comment by luisjimenez6245** [comment]
I can reproduce the same error with ormar 0.10.15 and 0.10.18

**Comment by luisjimenez6245** [comment]
After some testing, I think that there's an issue of a circular call of this functions or something like that, the solution that I have on this moment is to create an instance to save relation_map on an static property of the class and it reduced the time by 1/3.


<img width="1440" alt="Captura de Pantalla 2021-09-06 a la(s) 17 17 54" src="https://user-images.githubusercontent.com/26035464/132263332-42333222-52f7-451f-88f0-98768e2df74b.png">



```python
        print(dict_instance)
        if include and isinstance(include, Set):
            include = translate_list_to_dict(include)
        if exclude and isinstance(exclude, Set):
            exclude = translate_list_to_dict(exclude)

        start_time = time()
        if(self.__class__.__relation_model__ is None):
            relation_map = (
                relation_map
                if relation_map is not None
                else translate_list_to_dict(self._iterate_related_models())
            )
            self.__class__.__relation_model__ = relation_map
        else:
            relation_map = self.__class__.__relation_model__
        procces_time = time() - start_time
        print(procces_time)
        print(relation_map)
```

**Comment by luisjimenez6245** [comment]
After making another call to the function im getting the same response on only 2.43 sec, so I think that my models are making a very big relation map.

**Comment by luisjimenez6245** [comment]
After more tests, I reduced the time to 11ms after the second call.

**Comment by luisjimenez6245** [comment]
Sorry for leaving to many comments, but im trying to solve the issue. I Founded that the func _iterate_related_models of class  RelationMixin is calling too many times, so I added a limit for the recursion. I really want to hear of you, if you consider an issue to put a recursion limit.

**Comment by collerek** [comment]
The whole point of relation map is to limit recurrency as otherwise you would have infinite loop between models. It seems i missed some kind of scenario, will check this out. The map itself should limit the recurrent calls.

**Comment by collerek** [comment]
Since the relation map iterates all relations tree do you mind sharing the remaining models (and other models if the remaining models have further relations)?

From the code above I miss: Tag, User, ChangeLog, Label, ProjectLabel and UserLabel

**Comment by luisjimenez6245** [comment]
sorry for the delay, but here you have all my models



[models.txt](https://github.com/collerek/ormar/files/7141118/models.txt)

**Comment by collerek** [comment]
Ok I managed to reproduce the slow results (although for sqlite it takes like 4s not 50(!) it's way too long).

Wrote a test based on your models and will try to optimize it.

```python
@pytest.mark.asyncio
async def test_relation_dict():
    async with database:
        tags = [
            {"id": 18,
             "name": "name-18",
             "ref": "ref-18"},
            {"id": 17,
             "name": "name-17",
             "ref": "ref-17"},
            {"id": 12,
             "name": "name-12",
             "ref": "ref-12"}
        ]
        payload = [
            {
                "id": 9,
                "title": "prueba-2321",
                "description": "\n<!--- start changelog ver.v.1.3.0.0 -->\n### [v.1.3.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.3.0.0 -->\n",
                "data": {}
            },
            {
                "id": 8,
                "title": "prueba-123-prod",
                "description": "\n<!--- start changelog ver.v.1.2.0.0 -->\n### [v.1.2.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.2.0.0 -->\n",
                "data": {}
            },
            {
                "id": 6,
                "title": "prueba-3-2",
                "description": "\n<!--- start changelog ver.v.1.1.0.0 -->\n### [v.1.1.0.0] - 2021-07-29\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.1.0.0 -->\n",
                "data": {}
            }
        ]
        saved_tags = []
        for tag in tags:
            saved_tags.append(await Tag(**tag).save())

        for ind, pay in enumerate(payload):
            await Release(**pay, tag=saved_tags[ind]).save()

        releases = await Release.objects.order_by(Release.id.desc()).all()
        dicts = [release.dict() for release in releases]

        result = [
            {
                "id": 9,
                "title": "prueba-2321",
                "description": "\n<!--- start changelog ver.v.1.3.0.0 -->\n### [v.1.3.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.3.0.0 -->\n",
                "data": {},
                "tag": {
                    "id": 18,
                    "taglabel": None,
                    "tagcommit": None,
                    "tagissue": None
                },
                "changelogs": []
            },
            {
                "id": 8,
                "title": "prueba-123-prod",
                "description": "\n<!--- start changelog ver.v.1.2.0.0 -->\n### [v.1.2.0.0] - 2021-08-19\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.2.0.0 -->\n",
                "data": {},
                "tag": {
                    "id": 17,
                    "taglabel": None,
                    "tagcommit": None,
                    "tagissue": None
                },
                "changelogs": []
            },
            {
                "id": 6,
                "title": "prueba-3-2",
                "description": "\n<!--- start changelog ver.v.1.1.0.0 -->\n### [v.1.1.0.0] - 2021-07-29\n#### Resolved Issues\n\n#### Task\n\n- Probar flujo de changelog Automatic  Jira: [SAN-86](https://htech.atlassian.net/browse/SAN-86)\n\n    Description: Se probara el flujo de changelog automatic.  \n\n    Changelog: Se agrega función para extraer texto del campo changelog del dashboard de Sanval y ponerlo directamente en el changelog.md del repositorio. \n\n\n            \n<!--- end changelog ver.v.1.1.0.0 -->\n",
                "data": {},
                "tag": {
                    "id": 12,
                    "taglabel": None,
                    "tagcommit": None,
                    "tagissue": None
                },
                "changelogs": []
            }
        ]

        assert dicts == result
```

**Comment by collerek** [comment]
> After making another call to the function I'm getting the same response on only 2.43 sec, so I think that my models are making a very big relation map.

Yep, due to tags linking many models, your relation map has... almost 2600 entries. 😱 

I will have to optimize the algorithm on this as right now it's iterating the same models many times when comes to them from different related models. 🤔 

In the meantime, I did what you tried to do - so cache the `relation_map` in each `Model` class.

Can you install `ormar` from the `add_index` branch (pip install git+https://github.com/collerek/ormar.git@add_index
) and let me know how that looks for you after the update? 

Already the first call should be quicker, and the next one should be __much__ quicker.

This is a nice optimization anyway, so thanks! 😃

**Comment by luisjimenez6245** [comment]
Saving the relation map, improves a lot the time of making the dicts, but im still having a 10 sec time on my first request when trying to build the relation map, what I can suggest is to have a flag to limit the recursion. Thats the only way I can resolve my issue on calling the dict() func. What do you think?

**Comment by luisjimenez6245** [comment]
WOOOW!!
It is awesome the new implimentation it is sooo fast!

**Comment by luisjimenez6245** [comment]
Thanks for your help @collerek  and thanks for making ormar!

**Comment by collerek** [comment]
Fixed in #346
